python/vast/utils/arrow.py

```diff
@@ -1,7 +1,72 @@
-import ipaddress
+import ipaddress as ip
+from typing import SupportsBytes
 
 import pyarrow as pa
 
+# Accepts a 128-bit buffer holding an IPv6 address and returns an IPv4 or IPv6
+# address.
+def unpack_ip(buffer: SupportsBytes) -> ip.IPv4Address | ip.IPv6Address:
+    num = int.from_bytes(buffer, byteorder="big")
+    # Convert IPv4 mapped addresses back to regular IPv4.
+    # https://tools.ietf.org/html/rfc4291#section-2.5.5.2
+    if (num >> 32) == 65535:
+        num = num - (65535 << 32)
+    return ip.ip_address(num)
+
+
+class IPAddressScalar(pa.ExtensionScalar):
+    def as_py(self) -> ip.IPv4Address | ip.IPv6Address:
+        return unpack_ip(self.value.as_py())
+
+
+class IPAddressType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.binary(16), "vast.address")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(self, storage_type, serialized):
+        return IPAddressType()
+
+    def __reduce__(self):
+        return IPAddressScalar, ()
+
+    def __arrow_ext_scalar_class__(self):
+        return IPAddressScalar
+
+
+class SubnetScalar(pa.ExtensionScalar):
+    def as_py(self) -> ip.IPv4Network | ip.IPv6Network:
+        buffer = self.value.as_py()
+        network = unpack_ip(buffer[0:16])
+        length = buffer[16]
+        return ip.ip_network((network, length), strict=False)
+
+
+class SubnetType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.binary(17), "vast.subnet")
```

Comment:
This should match its equivalent in libvast:

https://github.com/tenzir/vast/blob/1c3eec37e96db254ad73a405164718b4d435d779/libvast/src/type.cpp#L1995-L2000

Reply:
I did this structurally. Unfortunately extension types don't compose in Python:

```python
        # This gives us a pyarrow.lib.ArrowNotImplementedError :-/
        # arrow_type = pa.struct([("address", AddressType()),
        #                       ("length", pa.uint8())])
        arrow_type = pa.struct([("address", pa.binary(16)), ("length", pa.uint8())])
```

Fortunately it doesn't matter in this circumstance.

Please resolve if you think this is adequately addressed. (Pun intended.)

Reply:
It doesn't matter when reading, but if we ever want to start generating arrays for our extension types inside Python (e.g., when modifying a column) then this will be a problem, because VAST won't be able to read them.

We should probably adapt the subnet extension type in VAST not to rely upon the address extension type then. We're not really using it in data, so it's only a theoretical and not a practical breaking change.

I'm fine with the change itself. Should we create a follow-up story for making the adjustment in VAST?

Reply:
Ah yes, good call. The interop is of course broken between C++ and Python if they don't use the *exact* same layout.

> I'm fine with the change itself. Should we create a follow-up story for making the adjustment in VAST?

I think we have to. 😒

Reply:
> I think we have to. 😒

Maybe not. I got a bit further in [96e3dfe](https://github.com/tenzir/vast/pull/2596/commits/96e3dfebf907974cfabc3f69d302effba7c1770a). Fortunately [`StructArray` instancess can have extension types](https://github.com/apache/arrow/blob/356e7f836c145966ebbeb65c3b65d82348e4234e/python/pyarrow/tests/test_extension_type.py#L825)!

There's just an issue when putting them into a `RecordBatch`. The error I'm getting now is:

```
Unsupported cast to extension<vast.address<AddressType>> from fixed_size_binary[16]
```

The folks at Huggingface have hand-rolled a [custom cast](https://github.com/huggingface/datasets/pull/3575). We may have to do the same here.

Reply:
I also created [ARROW-17839](https://issues.apache.org/jira/browse/ARROW-17839) for this.

Reply:
I've escalated this to [ARROW-17839](https://issues.apache.org/jira/browse/ARROW-17839).

Reply:
Fortunately there is a fix underway: https://github.com/apache/arrow/pull/14106.

Reply:
I kept disabled the failing unit test because it will likely take a a while until we can consume the fix here.

Since subnet types are quite rare in VAST data, I'm proposing that we proceed with this PR. The majority of Python analysis still works, and users will only be impacted when exporting subnet data. @dominiklohmann if this sounds acceptable, I would create a separate story for re-enabling the disabled test.

Reply:
That sounds like the right approach to me.